detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/GCPResource.java

```diff
@@ -0,0 +1,307 @@
+   *     necessary keys.
+   * @return a boolean indicating if the environment was determined to be GKE and GKE specific
+   *     attributes were applied.
+   */
+  private boolean generateGKEAttributesIfApplicable(AttributesBuilder attrBuilder) {
+    if (envVars.get("KUBERNETES_SERVICE_HOST") != null) {
+      attrBuilder.put(
+          ResourceAttributes.CLOUD_PLATFORM,
+          ResourceAttributes.CloudPlatformValues.GCP_KUBERNETES_ENGINE);
+      String podName = envVars.get("POD_NAME");
+      if (podName != null && !podName.isEmpty()) {
+        attrBuilder.put(ResourceAttributes.K8S_POD_NAME, podName);
+      } else {
+        // If nothing else is set, at least use hostname for pod name.
+        attrBuilder.put(ResourceAttributes.K8S_POD_NAME, envVars.get("HOSTNAME"));
+      }
+
+      String namespace = envVars.get("NAMESPACE");
+      if (namespace != null && !namespace.isEmpty()) {
+        attrBuilder.put(ResourceAttributes.K8S_NAMESPACE_NAME, namespace);
+      }
+
+      String containerName = envVars.get("CONTAINER_NAME");
+      if (containerName != null && !containerName.isEmpty()) {
+        attrBuilder.put(ResourceAttributes.K8S_CONTAINER_NAME, containerName);
+      }
+
+      String instanceId = metadata.getInstanceId();
+      if (instanceId != null) {
+        attrBuilder.put(ResourceAttributes.HOST_ID, instanceId);
+      }
+
+      String clusterLocation = metadata.getClusterLocation();
+      assignGKEAvailabilityZoneOrRegion(clusterLocation, attrBuilder);
+
+      String clusterName = metadata.getClusterName();
+      if (clusterName != null && !clusterName.isEmpty()) {
+        attrBuilder.put(ResourceAttributes.K8S_CLUSTER_NAME, clusterName);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Function that assigns either the cloud region or cloud availability zone depending on whether
+   * the cluster is regional or zonal respectively. Assigns both values if the cluster location
+   * passed is in an unexpected format.
+   *
+   * @param clusterLocation The location of the GKE cluster. Can either be an availability zone or a
+   *     region.
+   * @param attributesBuilder The {@link AttributesBuilder} object that needs to be updated with the
+   *     necessary keys.
+   */
+  private void assignGKEAvailabilityZoneOrRegion(
+      String clusterLocation, AttributesBuilder attributesBuilder) {
+    long dashCount =
+        StringUtils.isNullOrEmpty(clusterLocation)
+            ? 0
+            : clusterLocation.chars().filter(ch -> ch == '-').count();
+    switch ((int) dashCount) {
+      case 1:
+        // this is a region
+        attributesBuilder.put(ResourceAttributes.CLOUD_REGION, clusterLocation);
+        break;
+      case 2:
+        // this is a zone
+        attributesBuilder.put(ResourceAttributes.CLOUD_AVAILABILITY_ZONE, clusterLocation);
+        break;
+      default:
+        attributesBuilder.put(ResourceAttributes.CLOUD_REGION, clusterLocation);
```

Comment:
Seems best not to add zone or region if they are malformed?

Reply:
Good point.